### PR TITLE
Allow more natural fusion notation

### LINKED ISSUE #141: Remove reduce argument from fusion.fuse
It could be directly fused with ordinary function-call notation.

### PR DESCRIPTION
**Pull request by asi1024**
Improved fusion notation by allowing direct reduction.  This PR resolves #141.

### Example

```
@cupy.fuse()
def sample2(x, y):
    return cupy.sum(x + y, axis=0) * 2

for xp in (numpy, cupy):
    a = xp.array([[3, 1, 4, 1, 5], [9, 2, 6, 5, 3]], 'f')
    b = xp.array([[2, 7, 1, 8, 2], [8, 1, 8, 2, 8]], 'f')
    print(sample2(a, b))
```

```
[ 44.  22.  38.  32.  36.]
[ 44.  22.  38.  32.  36.]
```

### CODE REVIEW

**Comment by asi1024** [comment]
Removed in-place return value feature.

> - In-place return value
> ```
> @cupy.fuse()
> def sample1(x, y):
>     return x + y
> 
> for xp in (numpy, cupy):
>     a = xp.array([3, 1, 4, 1, 5], 'f')
>     b = xp.array([2, 7, 1, 8, 2], 'f')
>     c = sample1(a, b, a)
>     print(a, b, c)
> ```

File: cupy/core/fusion.py [hunk 1]
```diff
@@ -829,29 +925,42 @@ def _create_ufunc(cupy_ufunc, numpy_ufunc):
 
 class reduction(object):
 
-    def __init__(self, cupy_op, numpy_op):
+    def __init__(self, cupy_op, numpy_op, raw):
```
Comment (niboshi) [inline cupy/core/fusion.py]: What's the requirements of `raw`?

File: cupy/core/fusion.py [hunk 2]
```diff
@@ -829,29 +925,42 @@ def _create_ufunc(cupy_ufunc, numpy_ufunc):
...
+            history = arg._history
+            for op in self._raw._ops:
+                if numpy.can_cast(dtype.type, op[0][0]):
+                    return_type = numpy.dtype(op[1][0])
+                    history.premap_ret = _normalize_arg(arg, history)
+                    return_var = history.fresh_postmap_param(return_type)
+                    history.reduce_op = op
+                    history.reduce_identity = self.identity
+                    history.reduce_kwargs = kwargs
+                    history.preamble_set.add(self._preamble)
```
Comment (niboshi) [inline cupy/core/fusion.py]: Can you avoid altering `history`'s attributes directly? How about introducing a utility method?

File: cupy/core/fusion.py [hunk 3]
```diff
@@ -829,29 +925,42 @@ def _create_ufunc(cupy_ufunc, numpy_ufunc):
...
     def __call__(self, *args, **kwargs):
+        arg = args[0]
+        if isinstance(arg, FusionVarPython):
+            if len(args) != 1:
+                raise Exception("Can't reduce a tuple")
+            dtype = arg.dtype
+            history = arg._history
+            for op in self._raw._ops:
+                if numpy.can_cast(dtype.type, op[0][0]):
+                    return_type = numpy.dtype(op[1][0])
```
Comment (niboshi) [inline cupy/core/fusion.py]: Can you improve readability of these indexing on `op`? (at least comments?)

File: cupy/core/fusion.py [hunk 4]
```diff
@@ -496,84 +625,72 @@ def _get_fix_code(data_type, fixed_type, operation):
...
-        out_params = '%s res' % post_out.ty
-        return core.ReductionKernel(in_params, out_params, operation,
-                                    reduce_code,
-                                    'res = _post_map(_post_fix(a))',
-                                    identity,
-                                    name=name,
-                                    reduce_type=post_type,
-                                    preamble=submodule_code)
+        reduce_code = history.reduce_op[2][1]
+        rtype = history.reduce_op[2][3]
```
Comment (niboshi) [inline cupy/core/fusion.py]: ditto. Can you improve readability of these indexing?

File: cupy/core/fusion.py [hunk 5]
```diff
@@ -496,84 +625,72 @@ def _get_fix_code(data_type, fixed_type, operation):
...
-        if not out_params:
-            in_params = ', '.join(_get_params(in_vars[:-1]))
-            out_params = ', '.join(_get_params([in_vars[-1]]))
-        submodules = _gather_submodules(op_list)
-        submodule_code = ''.join(_get_submodule_code(_)
-                                 for _ in submodules.values())
-        return core.ElementwiseKernel(in_params, out_params,
-                                      operation, preamble=submodule_code,
-                                      name=name)
+def _get_fusion_from_types(func, in_types, name):
```
Comment (niboshi) [inline cupy/core/fusion.py]: Write a few lines of comments explaining what this function does.
Especially, what's the meaning of the return values?

File: cupy/core/fusion.py [hunk 6]
```diff
@@ -17,72 +17,284 @@
...
+
+    def _set_postmap_op(self, *args, **kwargs):
+        op = FusionOp(len(self.postmap_op_list), *args, **kwargs)
+        subm = op.func
+        self.submodules[subm.key()] = subm
+        self.postmap_op_list.append(op)
+        self.preamble_set.add(subm.preamble)
+        return op
+
+    def set_op(self, *args, **kwargs):
```
Comment (niboshi) [inline cupy/core/fusion.py]: `add_op` is better?

File: cupy/core/fusion.py [hunk 7]
```diff
@@ -829,29 +925,42 @@ def _create_ufunc(cupy_ufunc, numpy_ufunc):
...
+                    history.premap_ret = _normalize_arg(arg, history)
+                    return_var = history.fresh_postmap_param(return_type)
+                    history.reduce_op = op
+                    history.reduce_identity = self.identity
+                    history.reduce_kwargs = kwargs
+                    history.preamble_set.add(self._preamble)
+                    return ReducedVarPython(return_var, history)
+            raise TypeError("Type is mismatched. {}(...), {}".format(
+                self._raw._ops.name, dtype.type))
         if builtins.any(type(_) == numpy.ndarray for _ in args):
```
Comment (niboshi) [inline cupy/core/fusion.py]: Continue with `elif`.

File: cupy/core/fusion.py [hunk 8]
```diff
@@ -239,42 +451,34 @@ def copy(self):
...
 
 
-_kind_score = {
-    'b': 0,
-    'u': 1,
-    'i': 1,
-    'f': 2,
-    'c': 3,
-}
+class ReducedVarPython(FusionVarPython):
```
Comment (niboshi) [inline cupy/core/fusion.py]: Do we need a separate class? How about distinguishing with an attribute?

File: cupy/core/fusion.py [hunk 9]
```diff
@@ -17,72 +17,284 @@
...
-                      nin, nout, in_vars, out_vars, types, num)
+            self.op_list, self.var_list)
+
+    def is_postmap(self):
+        return self.reduce_op is not None
+
+    def ndarray_type(self):
+        return ReducedVarPython if self.is_postmap() else FusionVarPython
+
+    def fresh_index(self):
```
[the remark this answers is not in the document]
Comment (asi1024) [reply cupy/core/fusion.py]: "Fresh variable" is a standard terminology at least.

File: cupy/core/fusion.py (hunk 5 above)
Comment (asi1024) [reply cupy/core/fusion.py]: Documented.

File: cupy/core/fusion.py (hunk 3 above)
Comment (asi1024) [reply cupy/core/fusion.py]: Avoided indexing.

File: cupy/core/fusion.py (hunk 2 above)
Comment (asi1024) [reply cupy/core/fusion.py]: Fixed.

File: cupy/core/fusion.py (hunk 7 above)
Comment (asi1024) [reply cupy/core/fusion.py]: Fixed.

File: cupy/core/fusion.py [hunk 10]
```diff
@@ -239,59 +442,43 @@ def copy(self):
...
-    numpy.dtype('int8'): 'signed char',
-    numpy.dtype('uint64'): 'unsigned long long',
-    numpy.dtype('uint32'): 'unsigned int',
-    numpy.dtype('uint16'): 'unsigned short',
-    numpy.dtype('uint8'): 'unsigned char',
-    numpy.dtype('bool'): 'bool',
-}
-
-_dtype_list = [numpy.dtype(_) for _ in '?bhilqBHILQefdFD']
+def _get_cuda_var(arg):
```
[the remark this answers is not in the document]
Comment (asi1024) [reply cupy/core/fusion.py]: Made this function a `history`'s member method.

File: cupy/core/fusion.py [hunk 11]
```diff
@@ -17,72 +17,275 @@
...
+    # shape (tuple of ints): !! not supported !!
+    # _var (_FusionVarCUDA)
 
-    def __init__(self, var, mem):
+    def __init__(self, var):
         self._var = var
-        self.dtype = var.ty
-        self._mem = mem
+        self.dtype = var.dtype
+        self._is_postmap = _thread_local.history.is_postmap()
```
[the remark this answers is not in the document]
Comment (asi1024) [reply cupy/core/fusion.py]: Fixed.

File: cupy/core/fusion.py [hunk 12]
```diff
@@ -239,59 +442,43 @@ def copy(self):
...
 
-def _normalize_arg(arg, mem):
+    Return value: _FusionVarCUDA
+    """
+    history = _thread_local.history
     arg_type = type(arg)
-    if arg_type is _FusionRef:
-        return arg._var
+    if arg_type is FusionVarPython:
+        if arg._is_postmap is history.is_postmap():
```
[the remark this answers is not in the document]
Comment (asi1024) [reply cupy/core/fusion.py]: Fixed.

File: cupy/core/fusion.py [hunk 13]
```diff
@@ -623,32 +726,24 @@ def _call(self, *args, **kwargs):
...
+                types_str = '.'.join(repr(type(_)) for _ in args)
+                message = 'Can\'t fuse \n {}({})'.format(self.name, types_str)
                 warnings.warn(message)
-            if self.reduce is None:
-                return self.func(*args)
-            elif axis is None:
-                return self.post_map(self.reduce(self.func(*args)))
             else:
-                return self.post_map(self.reduce(self.func(*args), axis=axis))
+                return (self.func, {})
```
[the remark this answers is not in the document]
Comment (asi1024) [reply cupy/core/fusion.py]: Fixed.

File: tests/cupy_tests/core_tests/test_fusion.py [hunk 14]
```diff
@@ -1203,9 +1206,9 @@ def test_reduce1(self, xp, dtype):
         b = xp.array([[2, 2, 3, 3], [2, 2, 3, 3]], dtype=dtype)
         c = xp.array([[2, 3, 2, 3], [2, 3, 2, 3]], dtype=dtype)
 
-        @cupy.fuse(reduce=cupy.sum)
+        @cupy.fuse()
         def g(x, y, z):
-            return x * y + z
+            return cupy.sum(x * y + z)
```
[the remark this answers is not in the document]
Comment (asi1024) [reply tests/cupy_tests/core_tests/test_fusion.py]: Fixed.

File: tests/cupy_tests/core_tests/test_fusion.py [hunk 15]
```diff
@@ -660,15 +651,13 @@ def f(*x):
     def check_reduce(self, func, n, reduce_f, gen, args=None):
         if args is None:
             args = ((),) * n
-        self._check_reduce(func, n, reduce_f, gen, args, True)
-        self._check_reduce(func, n, reduce_f, gen, args, False)
-
-    def _check_reduce(self, func, n, reduce_f, gen, args, omit_nin):
-        nin = n if not omit_nin else None
+        self._check_reduce(func, n, reduce_f, gen, args)
+        self._check_reduce(func, n, reduce_f, gen, args)
```
[the remark this answers is not in the document]
Comment (asi1024) [reply tests/cupy_tests/core_tests/test_fusion.py]: Fixed.

File: cupy/core/fusion.py [hunk 16]
```diff
@@ -340,240 +522,168 @@ def res(*args, **kwargs):
...
-                                    preamble=submodule_code)
+        op = history.reduce_op
+        _, (fixed_type,), (_, reduce_code, fix_code, reduce_ctype) = op
+        if reduce_ctype is None:
+            reduce_ctype = 'type_in0_raw'
+
+        fixed_dtype = numpy.dtype(fixed_type)
+        fixed_ctype = _dtype_to_ctype[fixed_dtype]
+
+        postmap = '// {} operations\n'.format(len(history.postmap_op_list))
```
[the remark this answers is not in the document]
Comment (asi1024) [reply cupy/core/fusion.py]: Fixed.

File: cupy/core/fusion.py [hunk 17]
```diff
@@ -340,240 +522,168 @@ def res(*args, **kwargs):
...
 
-def _get_reduce_op(ops, dtype):
-    for i in ops._ops:
-        if numpy.can_cast(dtype.type, i[0][0]):
-            return i
-    raise TypeError("Type is mismatched. %s(...), %s" % (ops.name, dtype.type))
-
-
-def _get_post_code(post_vars, operation, post_out):
+def _postmap_code(in_param, out_params, operation):
```
[the remark this answers is not in the document]
Comment (asi1024) [reply cupy/core/fusion.py]: Fixed.

File: cupy/core/fusion.py [hunk 18]
```diff
@@ -828,30 +922,51 @@ def _create_ufunc(cupy_ufunc, numpy_ufunc):
...
+            for op in self._raw._ops:
+                (input_type,), (output_type,), _ = op
+                if numpy.can_cast(dtype.type, input_type):
+                    return_dtype = numpy.dtype(output_type)
+                    history.premap_ret = _get_cuda_var(arg)
+                    return_var = history.fresh_postmap_param(return_dtype)
+                    history.reduce_op = op
+                    history.reduce_identity = self.identity
+                    history.reduce_kwargs = kwargs
+                    history.add_preamble(self._preamble)
```
[the remark this answers is not in the document]
Comment (asi1024) [reply cupy/core/fusion.py]: Fixed.

File: cupy/core/fusion.py [hunk 19]
```diff
@@ -828,30 +922,51 @@ def _create_ufunc(cupy_ufunc, numpy_ufunc):
...
+        self.identity = raw.identity
+        self._preamble = raw._preamble
 
     def __call__(self, *args, **kwargs):
-        if builtins.any(type(_) == numpy.ndarray for _ in args):
+        arg = args[0]
+        if isinstance(arg, FusionVarPython):
+            if arg._is_postmap:
+                # Multiple reduction
+                raise NotImplementedError()
```
[the remark this answers is not in the document]
Comment (asi1024) [reply cupy/core/fusion.py]: Fixed.

File: cupy/core/fusion.py [hunk 20]
```diff
@@ -828,30 +922,51 @@ def _create_ufunc(cupy_ufunc, numpy_ufunc):
...
+            if len(args) != 1:
+                raise Exception('Can\'t reduce a tuple')
```
[the remark this answers is not in the document]
Comment (asi1024) [reply cupy/core/fusion.py]: Fixed.

File: cupy/core/fusion.py [hunk 21]
```diff
@@ -340,240 +522,168 @@ def res(*args, **kwargs):
...
+
+        postmap = '// {} operations\n'.format(len(history.postmap_op_list))
+        postmap += ''.join(v.declaration() for v in history.postmap_local_list)
+        postmap += ''.join(op.declaration_args()
+                           for op in history.postmap_op_list)
+        postmap += ''.join(op.code() for op in history.postmap_op_list)
+        postmap += ' '.join('{} = {};'.format(t, s)
+                            for s, t in zip(out_cvars, out_params))
+
+        submodules += history.get_submodules()
```
[the remark this answers is not in the document]
Comment (asi1024) [reply cupy/core/fusion.py]: Fixed.

File: cupy/core/fusion.py [hunk 22]
```diff
@@ -239,99 +344,192 @@ def copy(self):
...
+        return op
+
+    def add_op(self, *args, **kwargs):
+        if self.has_reduction():
+            return self._add_postmap_op(*args, **kwargs)
+        else:
+            return self._add_premap_op(*args, **kwargs)
+
+    def set_reduce_op(self, raw, arg, kwargs):
+        for op in raw._ops:
```
Comment (niboshi) [inline cupy/core/fusion.py]: How about `assert self.reduce_op is None`?

File: cupy/core/fusion.py [hunk 23]
```diff
@@ -239,99 +344,192 @@ def copy(self):
...
+            (input_type,), (output_type,), _ = op
+            if numpy.can_cast(arg.dtype.type, input_type):
+                return_dtype = numpy.dtype(output_type)
+                self.premap_ret = self.get_cuda_var(arg)
+                self.reduce_op = op
+                self.reduce_identity = raw.identity
+                self.reduce_kwargs = kwargs
+                self.add_preamble(raw._preamble)
+                return self.fresh_postmap_param(return_dtype)
+        return None
```
Comment (niboshi) [inline cupy/core/fusion.py]: How about raising `TypeError('Type is mismatched...` here?

File: cupy/core/fusion.py [hunk 24]
```diff
@@ -239,99 +344,192 @@ def copy(self):
...
+
+    def fresh_premap_param(self, *args, **kwargs):
+        index = self.fresh_index()
+        var = _FusionVarCUDA(index, *args, **kwargs)
+        self.param_list.append(var)
+        return var
+
+    def fresh_postmap_param(self, *args, **kwargs):
+        if self.postmap_param is not None:
+            raise NotImplementedError(
```
Comment (niboshi) [inline cupy/core/fusion.py]: I seems this case never happens?
If so, use `assert` instead of an exception.

File: cupy/core/fusion.py [hunk 25]
```diff
@@ -239,99 +344,192 @@ def copy(self):
...
+        self.count += 1
+        return res
+
+    def fresh_premap_param(self, *args, **kwargs):
+        index = self.fresh_index()
+        var = _FusionVarCUDA(index, *args, **kwargs)
+        self.param_list.append(var)
+        return var
+
+    def fresh_postmap_param(self, *args, **kwargs):
```
Comment (niboshi) [inline cupy/core/fusion.py]: Prefix with `_` if this method is used only within the class (and not meant to be used from outside).

File: cupy/core/fusion.py [hunk 26]
```diff
@@ -239,99 +344,192 @@ def copy(self):
...
+        is_ndarray = hasattr(arg, 'dtype') and arg.dtype in _dtype_list
+        if is_scalar or is_ndarray:
+            return self.fresh_local(numpy.dtype(arg_type), const=arg)
+        raise Exception('Unsupported type {}'.format(arg_type))
+
+    def call_ufunc(self, ufunc, args, kwargs):
+        nin = ufunc.nin
+        nout = ufunc.nout
+
+        def _should_use_min_scalar(in_args):
```
Comment (niboshi) [inline cupy/core/fusion.py]: Please write the brief description of `_should_use_min_scalar`, `can_cast1` and `can_cast2`, respectively.

File: cupy/core/fusion.py [hunk 27]
```diff
@@ -239,99 +344,192 @@ def copy(self):
...
+            return True
 
-    def res(*args, **kwargs):
-        mem = get_mem(args)
-        var_list = [_normalize_arg(_, mem) for _ in args]
+        var_list = [self.get_cuda_var(_) for _ in args]
         if 'out' in kwargs:
-            var_list.append(_normalize_arg(kwargs.pop('out'), mem))
+            var = self.get_cuda_var(kwargs.pop('out'))
+            var_list.append(var)
```
Comment (niboshi) [inline cupy/core/fusion.py]: It seems you don't have to assign a variable `var`.

File: cupy/core/fusion.py [hunk 28]
```diff
@@ -340,240 +538,186 @@ def res(*args, **kwargs):
             ty_ins = [numpy.dtype(_) for _ in ty_ins]
             ty_outs = [numpy.dtype(_) for _ in ty_outs]
```
Comment (niboshi) [inline cupy/core/fusion.py]: How about `ty_ins/outs` -> `dtypes_in` and `dtypes_out` respectively (after converted to dtypes)?

File: cupy/core/fusion.py [hunk 29]
```diff
@@ -340,240 +538,186 @@ def res(*args, **kwargs):
...
+        ${operation};
+        return out0;
+        }
+        ''').substitute(
+            reduce_ctype=reduce_ctype,
+            fixed_ctype=_dtype_to_ctype[fixed_dtype],
+            operation=operation)
+        return module_code
+
+    def get_fusion_from_types(self, func, in_dtypes, name):
```
Comment (niboshi) [inline cupy/core/fusion.py]: Is it meaningful to emphasize `from_types` in the method name? It seems not to me.

File: cupy/core/fusion.py [hunk 30]
```diff
@@ -340,240 +538,186 @@ def res(*args, **kwargs):
...
-                                    name=name,
-                                    reduce_type=post_type,
-                                    preamble=submodule_code)
+            ufunc.name, in_dtypes, out_dtypes))
+
+    def call_elementwise(self, f, args, kwargs):
+        raise NotImplementedError(
+            'Fusion for elementwise-kernel is not implemented yet')
+
+    def call(self, f, args, kwargs):
```
Comment (niboshi) [inline cupy/core/fusion.py]: From where this method is called? Seems unused?

File: cupy/core/fusion.py [hunk 31]
```diff
@@ -340,240 +538,186 @@ def res(*args, **kwargs):
...
+            postmap_code += ''.join(op.code() for op in self.postmap_op_list)
+            postmap_code += ' '.join('{} = {};'.format(t, s)
+                                     for s, t in zip(out_cvars, out_params))
+
+            submodule_code += self.emit_premap_code(in_params, operation)
+            submodule_code += 'typedef {} type_in0_raw;\n'.format(fixed_ctype)
+            submodule_code += 'typedef {} type_out0_raw;\n'.format(fixed_ctype)
+            submodule_code += self.emit_fix_code(
+                reduce_ctype, fixed_dtype, fix_code)
+            submodule_code += self.emit_postmap_code(out_params, postmap_code)
```
Comment (niboshi) [inline cupy/core/fusion.py]: What is "fix code"?
Is "fix" related to "fixed" in `fixed_dtype` etc.?
Please write somewhere.

File: cupy/core/fusion.py [hunk 32]
```diff
@@ -239,99 +344,192 @@ def copy(self):
...
+    def get_cuda_var(self, arg):
+        """This converts `arg` to _FusionVarCUDA data.
+
+        Args:
+            arg (FusionVarPython or a primitive type)
+
+        Return value: _FusionVarCUDA
+        """
+        arg_type = type(arg)
+        if arg_type is FusionVarPython:
```
Comment (niboshi) [inline cupy/core/fusion.py]: Use `isinstance`, unless there's any reason (such as performance).

File: cupy/core/fusion.py [hunk 33]
```diff
@@ -239,99 +344,192 @@ def copy(self):
...
+            if arg._is_postmap == self.has_reduction():
+                return arg._var
             else:
-                if not numpy.can_cast(args[i].const, ty_ins[i]):
-                    return False
-        return True
+                # Map operation between pre-map variable and post-map variable
+                raise Exception('Shape mismatch')
+        is_scalar = arg_type in six.integer_types + (float, bool, complex)
```
Comment (niboshi) [inline cupy/core/fusion.py]: Ditto. Use `isinstance`.

File: cupy/core/fusion.py [hunk 34]
```diff
@@ -623,32 +760,24 @@ def _call(self, *args, **kwargs):
...
-                return self.func(*args)
-            elif axis is None:
-                return self.post_map(self.reduce(self.func(*args)))
             else:
-                return self.post_map(self.reduce(self.func(*args), axis=axis))
+                return self.func, {}
+
+    def _call(self, *args, **kwargs):
+        func, kw = self.compile(*args, **kwargs)
+        kwargs = dict(six.itertools.chain(kwargs.items(), kw.items()))
```
Comment (niboshi) [inline cupy/core/fusion.py]: How about `dict(kwargs, **kw)`?

File: cupy/core/fusion.py [hunk 35]
```diff
@@ -702,15 +829,17 @@ def __init__(self, fusion_op, cupy_op, numpy_op):
...
         return repr(self._cupy_op)
 
     def __call__(self, *args, **kwargs):
         in_fusion = getattr(_thread_local, 'in_fusion', False)
         if in_fusion:
-            if builtins.any(isinstance(_, _FusionRef) for _ in args):
-                return _convert(self._fusion_op)(*args, **kwargs)
+            if builtins.any(isinstance(_, FusionVarPython) for _ in args):
+                history = _thread_local.history
+                return history.call_ufunc(self._fusion_op, args, kwargs)
```
Comment (niboshi) [inline cupy/core/fusion.py]: This assignment of variable `history` is not needed.

File: cupy/core/fusion.py (hunk 26 above)
Comment (asi1024) [reply cupy/core/fusion.py]: Added a brief note.

**Comment by asi1024** [comment]
This PR also resolves #1021.

**Comment by niboshi** [comment]
LGTM!

**Comment by okuta** [comment]
Can I backport this?

**Pull request merged**